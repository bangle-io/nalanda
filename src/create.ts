// 3 Generics that are optional

import type { SliceConfig } from './slice';
import { Slice, SliceKey } from './slice';
import type { EffectsBase, RawAction, SelectorFn } from './types';

export function createKey<SS extends object>(initState: SS) {
  function _createKey<K extends string, DS extends Slice[]>(
    id: K,
    deps: DS,
  ): SliceKey<K, SS, {}, DS>;
  function _createKey<
    K extends string,
    DS extends Slice[],
    SE extends Record<string, SelectorFn<SS, DS, any>>,
  >(id: K, deps: DS, selector: SE): SliceKey<K, SS, {}, DS>;
  function _createKey(id: any, deps: any, selector?: any): any {
    return new SliceKey(id, deps, initState, selector || {});
  }

  return _createKey;
}

export function key<K extends string, SS extends object, DS extends Slice[]>(
  id: K,
  deps: DS,
  initState: SS,
): SliceKey<K, SS, {}, DS>;
export function key<
  K extends string,
  SS extends object,
  DS extends Slice[],
  SE extends Record<string, SelectorFn<SS, DS, any>>,
>(id: K, deps: DS, initState: SS, selector: SE): SliceKey<K, SS, SE, DS>;
export function key(id: any, deps: any, initState: any, selector?: any): any {
  return new SliceKey(id, deps, initState, selector || {});
}

export function slice<
  SK extends SliceKey<any, any, any, any>,
  A extends Record<
    string,
    RawAction<any[], SK['initState'], SK['dependencies']>
  >,
>({
  key,
  actions,
  effects,
  config = {},
}: {
  key: SK;
  actions: A;
  effects?: EffectsBase<
    Slice<SK['key'], SK['initState'], SK['dependencies'], A, SK['selectors']>
  >;
  config?: SliceConfig;
}): Slice<SK['key'], SK['initState'], SK['dependencies'], A, SK['selectors']> {
  return new Slice(key, actions, effects ? [effects] : [], config);
}
