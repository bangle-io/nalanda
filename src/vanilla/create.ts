import { mapObjectValues } from './helpers';
import {
  createSliceKey,
  createSliceNameOpaque,
  LineageId,
  VoidFn,
} from './internal-types';
import {
  ActionBuilder,
  AnySlice,
  Effect,
  SelectorFn,
  TxCreator,
} from './public-types';
import { Slice } from './slice';
import { Transaction } from './transaction';

class SliceKey<
  N extends string,
  SS extends object,
  SE extends SelectorFn<SS, DS, any>,
  DS extends AnySlice,
> {
  constructor(
    public name: N,
    public dependencies: DS[],
    public initState: SS,
    public selector: SE,
  ) {}
}

export function createKey<
  K extends string,
  SS extends object,
  DS extends AnySlice,
>(id: K, deps: DS[], initState: SS): SliceKey<K, SS, VoidFn, DS>;
export function createKey<
  K extends string,
  SS extends object,
  DS extends AnySlice,
  SE extends SelectorFn<SS, DS, any>,
>(id: K, deps: DS[], initState: SS, selector: SE): SliceKey<K, SS, SE, DS>;
export function createKey(
  id: any,
  deps: any[],
  initState: any,
  selector?: any,
): any {
  return new SliceKey(id, deps, initState, selector || (() => {}));
}

type InferName<SK extends SliceKey<any, any, any, any>> = SK extends SliceKey<
  infer N,
  any,
  any,
  any
>
  ? N
  : never;
type InferInitState<SK extends SliceKey<any, any, any, any>> =
  SK extends SliceKey<any, infer SS, any, any> ? SS : never;
type InferDependencies<SK extends SliceKey<any, any, any, any>> =
  SK extends SliceKey<any, any, any, infer DS> ? DS : never;
type InferSelector<SK extends SliceKey<any, any, any, any>> =
  SK extends SliceKey<any, any, infer SE, any> ? SE : never;

export function slice<
  SK extends SliceKey<any, any, any, any>,
  A extends Record<
    string,
    ActionBuilder<any[], InferInitState<SK>, InferDependencies<SK>>
  >,
>({
  key,
  actions,
  effects,
}: {
  key: SK;
  actions: A;
  effects?: Effect<
    SK['name'],
    InferInitState<SK>,
    InferDependencies<SK>,
    any,
    InferSelector<SK>
  >[];
}): Slice<
  InferName<SK>,
  InferInitState<SK>,
  InferDependencies<SK>,
  ActionBuilderRecordConvert<InferName<SK>, A>,
  InferSelector<SK>
> {
  let sel: typeof key.selector = key.selector || (() => {});
  const slice = new Slice({
    actions: ({ lineageId }) =>
      expandActionBuilders(key.name, actions, lineageId),
    reducer: (sliceState, storeState, tx) => {
      const apply = actions[tx.actionId];

      if (!apply) {
        throw new Error(
          `Action "${tx.actionId}" not found in Slice "${key.name}"`,
        );
      }
      return apply(...tx.payload)(sliceState, storeState);
    },
    dependencies: key.dependencies,
    effects: effects || [],
    initState: key.initState,
    name: key.name,
    selector: sel,
  });

  return slice;
}

type ActionBuilderRecordConvert<
  K extends string,
  A extends Record<string, any>,
> = {
  [KK in keyof A]: A[KK] extends ActionBuilder<infer P, any, any>
    ? TxCreator<K, P>
    : never;
};

export function createSlice<
  N extends string,
  SS extends object,
  DS extends Slice<string, any, any, {}, VoidFn>,
  A extends Record<string, ActionBuilder<any[], SS, DS>>,
  SE extends SelectorFn<SS, DS, any>,
>(
  dependencies: DS[],
  arg: {
    name: N;
    initState: SS;
    actions: A;
    selector: SE;
    terminal?: boolean;
  },
): Slice<N, SS, DS, ActionBuilderRecordConvert<N, A>, SE> {
  const slice = new Slice({
    actions: ({ lineageId }) =>
      expandActionBuilders(arg.name, arg.actions, lineageId),
    dependencies,
    effects: [],
    initState: arg.initState,
    name: arg.name,
    selector: arg.selector || (() => {}),
    reducer: (sliceState, storeState, tx) => {
      const apply = arg.actions[tx.actionId];

      if (!apply) {
        throw new Error(
          `Action "${tx.actionId}" not found in Slice "${arg.name}"`,
        );
      }

      return apply(...tx.payload)(sliceState, storeState);
    },
    terminal: arg.terminal || false,
  });

  return slice;
}

function expandActionBuilders<
  N extends string,
  A extends Record<string, ActionBuilder<any[], any, any>>,
>(name: N, actions: A, lineageId: LineageId): ActionBuilderRecordConvert<N, A> {
  let sliceName = createSliceNameOpaque(name);
  const result: Record<string, TxCreator> = mapObjectValues(
    actions,
    (action, actionId): TxCreator => {
      return (...params) => {
        return new Transaction({
          sourceSliceName: sliceName,
          targetSliceLineage: lineageId,
          payload: params,
          actionId,
        });
      };
    },
  );

  return result as any;
}
