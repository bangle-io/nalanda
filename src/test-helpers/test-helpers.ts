import { Transaction } from '../vanilla';
import { AnySlice } from '../vanilla/slice';
import { DispatchTx, ReducedStore, Store } from '../vanilla/store';
import { LogItem, TX_META_DISPATCHER } from '../vanilla/transaction';

export function testOverrideDependencies<SL extends AnySlice>(
  slice: SL,
  dependencies: AnySlice[] = slice.spec.dependencies,
  cloneSpec = false,
): SL {
  let spec = slice.spec;
  if (cloneSpec) {
    spec = { ...spec };
  }
  (spec as any).dependencies = dependencies;

  (slice as any).spec = spec;

  return slice;
}

export function waitUntil<TStore extends ReducedStore<any> | Store<any>>(
  store: TStore,
  condition: (state: TStore['state']) => boolean,
  waitUntil = 100,
  pollFrequency = 5,
): Promise<TStore['state']> {
  let interval: ReturnType<typeof setInterval> | undefined;
  let timeout: ReturnType<typeof setTimeout> | undefined;

  return new Promise<TStore['state']>((resolve, reject) => {
    timeout = setTimeout(() => {
      clearInterval(interval);
      reject(new Error('Timeout condition not met'));
    }, waitUntil);

    interval = setInterval(() => {
      if (condition(store.state)) {
        clearTimeout(timeout);
        clearInterval(interval);
        resolve(store.state);
      }
    }, pollFrequency);
  });
}

export function createDispatchSpy(fn?: (tx: Transaction<any, any>) => void) {
  let txs: Transaction<any, any>[] = [];
  const dispatch: DispatchTx<Transaction<any, any>> = (store, tx) => {
    let newState = store.state.applyTransaction(tx);
    fn?.(tx);
    txs.push(tx);
    Store.updateState(store, newState, tx);
  };
  let logItems: LogItem[] = [];
  return {
    debug: (log: LogItem) => {
      if (log.type === 'TX') {
        logItems.push({ ...log, txId: '<txId>' });
        return;
      }
      logItems.push(log);
    },
    getDebugLogItems() {
      return logItems;
    },
    dispatch,
    getTransactions() {
      return txs;
    },
    getSimplifiedTransactions({
      filterBySource,
    }: {
      hideInternal?: boolean;
      filterBySource?: AnySlice | AnySlice[];
    } = {}) {
      return txs

        .filter((r) => {
          return !filterBySource
            ? true
            : [filterBySource]
                .flatMap((s) => s)
                .some((s) => r.sourceSliceLineage == s.spec.lineageId);
        })
        .map(
          ({
            sourceSliceLineage,
            metadata,
            payload,
            config,
            targetSliceLineage,
          }) => ({
            sourceSliceLineage,
            actionId: config.actionId,
            payload,
            dispatchSource: metadata.getMetadata(TX_META_DISPATCHER),
            targetSliceLineage,
          }),
        );
    },
  };
}
