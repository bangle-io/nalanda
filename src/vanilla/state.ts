import { findDuplications } from './helpers';
import { KeyMapping } from './merge';
import { BareSlice } from './slice';
import { Transaction } from './transaction';

export type ResolveSliceIfRegistered<
  SL extends BareSlice,
  SliceRegistry extends BareSlice,
> = SL extends BareSlice<infer K, any>
  ? K extends SliceRegistry['key']
    ? SL
    : never
  : never;

export interface StoreState<RegSlices extends BareSlice> {
  getSliceState<SL extends BareSlice>(
    slice: ResolveSliceIfRegistered<SL, RegSlices>,
  ): SL['initState'];

  applyTransaction(
    tx: Transaction<RegSlices['key'], unknown[]>,
  ): StoreState<RegSlices>;
}

interface StoreStateOptions {
  debug?: boolean;
  keyMapping?: KeyMapping | undefined;
}

export class InternalStoreState implements StoreState<any> {
  protected slicesCurrentState: Record<string, unknown> = Object.create(null);

  static create<SL extends BareSlice>(slices: SL[]): StoreState<SL> {
    InternalStoreState.checkUniqueKeys(slices);
    InternalStoreState.circularCheck(slices);
    InternalStoreState.checkDependencyOrder(slices);

    const instance = new InternalStoreState(slices);

    for (const slice of slices) {
      instance.slicesCurrentState[slice.key] = slice.initState;
    }

    return instance;
  }

  constructor(
    public readonly _slices: BareSlice[],
    public opts?: StoreStateOptions,
  ) {}

  applyTransaction(tx: Transaction<string, unknown[]>): InternalStoreState {
    const newState = { ...this.slicesCurrentState };
    const newStoreState = this._fork(newState);

    let found = false;

    for (const slice of this._slices) {
      if (slice.key === tx.sliceKey) {
        found = true;

        const sliceState = newStoreState._getDirectSliceState(slice.key);

        if (!sliceState.found) {
          throw new Error(
            `Slice "${slice.key}" or one of its dependencies not found in store`,
          );
        }

        const scopedStoreState = newStoreState._withKeyMapping(
          slice._bare.keyMapping,
        );

        newState[slice.key] = slice.applyTx(
          sliceState.value,
          scopedStoreState,
          tx,
        );
      }
    }

    if (!found) {
      return this;
    }

    return newStoreState;
  }

  getSliceState(sl: BareSlice): unknown {
    debugger;
    let result = this._getDirectSliceState(sl.key);
    if (!result.found) {
      throw new Error(`Slice "${sl.key}" not found in store`);
    }
    return result.value;
  }

  private _getDirectSliceState(key: string) {
    if (this.opts?.keyMapping) {
      const mappedKey = this.opts.keyMapping.get(key);
      if (mappedKey === undefined) {
        throw new Error(
          `Key "${key}" not found in keyMapping. Did you forget to add it as a dependency it?`,
        );
      }
      console.debug(`Augmented key "${key}" to "${mappedKey}`);
      key = mappedKey;
    }

    if (Object.prototype.hasOwnProperty.call(this.slicesCurrentState, key)) {
      return {
        found: true,
        value: this.slicesCurrentState[key]!,
      };
    }

    return { found: false, value: undefined };
  }

  _withKeyMapping(keyMapping?: KeyMapping) {
    return this._fork(this.slicesCurrentState, { keyMapping });
  }

  private _fork(
    slicesState: Record<string, unknown>,
    opts?: Partial<StoreStateOptions>,
  ): InternalStoreState {
    const newOpts = !opts
      ? this.opts
      : {
          ...this.opts,
          ...opts,
        };

    const newInstance = new InternalStoreState(this._slices, newOpts);
    newInstance.slicesCurrentState = slicesState;
    return newInstance;
  }

  static checkDependencyOrder(slices: BareSlice[]) {
    let seenKeys = new Set<string>();
    for (const slice of slices) {
      const { dependencies } = slice.config;

      if (dependencies !== undefined) {
        const depKeys = dependencies.map((d) => d.key);

        for (const depKey of depKeys) {
          if (!seenKeys.has(depKey)) {
            throw new Error(
              `Slice "${slice.key}" has a dependency on Slice "${depKey}" which is either not registered or is registered after this slice.`,
            );
          }
        }
      }
      seenKeys.add(slice.key);
    }
  }

  static checkUniqueKeys(slices: BareSlice[]) {
    const keys = slices.map((s) => s.key);
    const unique = new Set(keys);

    if (keys.length !== unique.size) {
      const dups = findDuplications(keys);
      throw new Error('Duplicate slice keys ' + dups.join(', '));
    }
  }

  static circularCheck(slices: BareSlice[]) {
    const stack = new Set<string>();
    const visited = new Set<string>();

    const checkCycle = (slice: BareSlice): boolean => {
      const key = slice.key;
      if (stack.has(key)) return true;
      if (visited.has(key)) return false;

      visited.add(key);
      stack.add(key);

      for (const dep of slice.config.dependencies) {
        if (checkCycle(dep)) {
          return true;
        }
      }
      stack.delete(key);
      return false;
    };

    for (const slice of slices) {
      const cycle = checkCycle(slice);
      if (cycle) {
        const path = [...stack];
        path.push(slice.key);

        throw new Error(
          `Circular dependency detected in slice "${
            slice.key
          }" with path ${path.join(' ->')}`,
        );
      }
    }
  }
}
