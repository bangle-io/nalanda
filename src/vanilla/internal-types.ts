import { BareSlice } from './slice';
import { StoreState } from './state';
import { Transaction } from './transaction';

export type IfAny<T, Y, N> = 0 extends 1 & T ? Y : N;

export type AnyFn = (...args: any[]) => any;

export const expectType = <Type>(_: Type): void => void 0;
export function rejectAny<K extends IfAny<K, never, unknown>>(key: K): void {}

export type TxApplicator<K extends string, SS = any> = (
  sliceState: SS,
  storeState: StoreState<BareSlice>,
  tx: Transaction<K, unknown[]>,
) => SS;

export type ExtractReturnTypes<
  T extends Record<string, (...args: any[]) => any>,
> = {
  [K in keyof T]: T[K] extends (i: any) => infer R ? R : never;
};
