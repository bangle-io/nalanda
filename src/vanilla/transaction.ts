import { uuid } from './helpers';
import { LineageId } from './internal-types';

const contextId = uuid(5);
let counter = 0;

function incrementalId() {
  return `tx_${contextId}-${counter++}`;
}

export const TX_META_DISPATCHER = 'TX_META_DISPATCHER';
export const TX_META_DISPATCH_INFO = 'TX_META_DISPATCH_INFO';
export const TX_META_STORE_NAME = 'TX_META_STORE_NAME';
export const TX_META_DESERIALIZED_FROM = 'TX_META_DESERIALIZED_FROM';
export const TX_META_DESERIALIZED_META = 'TX_META_DESERIALIZED_META';
export const TX_META_CHANGE_LINEAGE = 'TX_META_CHANGE_LINEAGE';

export class Transaction<N extends string, P extends unknown[]> {
  public metadata = new Metadata();

  public readonly sourceSliceLineage: LineageId;
  public readonly targetSliceLineage: LineageId;
  public readonly payload: P;
  public readonly actionId: string;
  public readonly uid = incrementalId();

  toJSONObj(payloadSerializer: (payload: unknown[]) => string) {
    return {
      sourceSliceLineage: this.sourceSliceLineage,
      targetSliceLineage: this.targetSliceLineage,
      sourceSliceName: this.config.sourceSliceName,
      payload: payloadSerializer(this.payload),
      actionId: this.actionId,
      uid: this.uid,
      metadata: this.metadata.toJSONObj(),
    };
  }

  static fromJSONObj(
    obj: ReturnType<Transaction<any, any>['toJSONObj']>,
    payloadParser: (payload: string) => unknown[],
    debugInfo?: string,
  ) {
    let tx = new Transaction({
      sourceSliceLineage: obj.sourceSliceLineage,
      targetSliceLineage: obj.targetSliceLineage,
      sourceSliceName: obj.sourceSliceName,
      payload: payloadParser(obj.payload),
      actionId: obj.actionId,
    });
    tx.metadata = Metadata.fromJSONObj(obj.metadata);
    tx.metadata.appendMetadata(TX_META_DESERIALIZED_FROM, obj.uid);

    if (debugInfo) {
      tx.metadata.appendMetadata(TX_META_DESERIALIZED_META, debugInfo);
    }

    return tx;
  }

  constructor(
    public readonly config: {
      sourceSliceLineage?: LineageId | undefined;
      sourceSliceName: N;
      targetSliceLineage: LineageId;
      payload: P;
      actionId: string;
    },
  ) {
    this.targetSliceLineage = config.targetSliceLineage;
    this.sourceSliceLineage =
      config.sourceSliceLineage || config.targetSliceLineage;
    this.payload = config.payload;
    this.actionId = config.actionId;
  }

  change({
    targetSliceLineage,
    sourceSliceLineage,
  }: {
    targetSliceLineage: LineageId;
    sourceSliceLineage?: LineageId;
  }): Transaction<N, P> {
    const tx = new Transaction({
      ...this.config,
      targetSliceLineage,
      sourceSliceLineage: sourceSliceLineage,
    });
    tx.metadata = this.metadata.fork();

    tx.metadata.appendMetadata(
      TX_META_CHANGE_LINEAGE,
      `target ${this.targetSliceLineage} -> ${targetSliceLineage}` +
        (sourceSliceLineage
          ? ` source ${this.sourceSliceLineage} -> ${sourceSliceLineage}`
          : ''),
    );

    return tx;
  }
}

export class Metadata {
  private _metadata: Record<string, string> = Object.create(null);

  static fromJSONObj(obj: Record<string, string>) {
    let meta = new Metadata();
    meta._metadata = { ...obj };
    return meta;
  }

  toJSONObj() {
    return { ...this._metadata };
  }

  appendMetadata(key: string, val: string) {
    let existing = this.getMetadata(key);
    this.setMetadata(key, existing ? existing + ',' + val : val);
  }

  getMetadata(key: string) {
    return this._metadata[key];
  }

  setMetadata(key: string, val: string) {
    this._metadata[key] = val;
  }

  fork(): Metadata {
    const meta = new Metadata();
    meta._metadata = { ...this._metadata };
    return meta;
  }
}

export interface EffectLog {
  type: 'SYNC_UPDATE_EFFECT' | 'UPDATE_EFFECT';
  name: string;
  source: Array<{ lineageId: LineageId; actionId: string }>;
}

export interface TransactionLog {
  type: 'TX';
  sourceSliceLineage: LineageId;
  targetSliceLineage: LineageId;
  actionId: string;
  dispatcher?: string;
  dispatchInfo?: string;
  store?: string;
  txId: string;
  payload: unknown[];
}

export function txLog(tx: Transaction<any, any>): TransactionLog {
  return Object.fromEntries(
    Object.entries({
      type: 'TX',
      sourceSliceLineage: tx.sourceSliceLineage,
      targetSliceLineage: tx.targetSliceLineage,
      actionId: tx.config.actionId,
      dispatcher: tx.metadata.getMetadata(TX_META_DISPATCHER),
      dispatchInfo: tx.metadata.getMetadata(TX_META_DISPATCH_INFO),
      store: tx.metadata.getMetadata(TX_META_STORE_NAME),
      txId: tx.uid,
      payload: tx.payload,
    }).filter((r) => r[1] !== undefined),
  ) as any;
}

export type LogItem = EffectLog | TransactionLog;

export type DebugFunc = (item: LogItem) => void;
