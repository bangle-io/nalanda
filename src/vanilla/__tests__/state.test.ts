import { testOverrideSlice } from '../../test-helpers';
import { createKey, slice } from '../create';
import { createSliceKey, expectType } from '../internal-types';
import { InternalStoreState } from '../state';
import { Transaction } from '../transaction';

const testSlice1 = slice({
  key: createKey('test-1', [], { num: 4 }),
  actions: {
    increment: (opts: { increment: boolean }) => (state) => {
      return { ...state, num: state.num + (opts.increment ? 1 : 0) };
    },
    decrement: (opts: { decrement: boolean }) => (state) => {
      return { ...state, num: state.num - (opts.decrement ? 1 : 0) };
    },
  },
});

const testSlice2 = slice({
  key: createKey('test-2', [], { name: 'tame' }),
  actions: {
    prefix: (prefix: string) => (state) => {
      return { ...state, name: prefix + state.name };
    },
    padEnd: (length: number, pad: string) => (state) => {
      return { ...state, name: state.name.padEnd(length, pad) };
    },
    uppercase: () => (state) => {
      return { ...state, name: state.name.toUpperCase() };
    },
  },
});

const depTestSlice1 = slice({
  key: createKey('dep-test-1', [testSlice1], { myDep: 4, myDepStr: 'hi' }),
  actions: {
    increment: () => (state, storeState) => ({
      ...state,
      myDep: state.myDep + 1 + testSlice1.getState(storeState).num,
    }),
  },
});

describe('applyTransaction', () => {
  test('applyTransaction works', () => {
    const mySlice = slice({
      key: createKey('test', [], { num: 1 }),
      actions: {
        myAction: (num: number) => (state) => {
          return { ...state, num: num + state.num };
        },
        action2: (num: number, foo: string, brother: () => void) => (state) =>
          state,
      },
    });

    const state = InternalStoreState.create([mySlice]);

    // @ts-expect-error - should error when a field is not defined
    let testVal0 = mySlice.actions.myAction(5).randomValue;

    expectType<number[]>(mySlice.actions.myAction(5).payload);
    expectType<[number, string, () => void]>(
      mySlice.actions.action2(5, 'str', () => {
        return;
      }).payload,
    );

    expect(mySlice.actions.myAction(5).payload).toEqual([5]);
    expect(
      mySlice.actions.action2(5, 'str', () => {
        return;
      }).payload,
    ).toEqual([5, 'str', expect.any(Function)]);

    const newState = state.applyTransaction(mySlice.actions.myAction(5))!;

    expect(newState.getSliceState(mySlice)).toEqual({
      num: 6,
    });
  });

  test('applyTransaction works with dependencies', () => {
    const sliceDep1 = slice({
      key: createKey('test-dep1', [], { num: 50 }),
      actions: {},
    });
    const sliceDep2 = slice({
      key: createKey('test-dep2', [], { num: 3 }),
      actions: {},
    });

    const mySlice = slice({
      key: createKey('test', [sliceDep1, sliceDep2], { num: 1 }),
      actions: {
        myAction: (num: number) => (state, storeState) => {
          // @ts-expect-error - should not allow access of any unknown field in the state
          let testVal1 = state.xyzWrong;

          let dep1 = sliceDep1.getState(storeState);

          // @ts-expect-error - should not allow access of any unknown field in the state
          let testVal2 = dep1.xyzWrong;

          let dep2 = sliceDep2.getState(storeState);

          return {
            ...state,
            num: num + state.num + dep1.num + dep2.num,
          };
        },
      },
    });

    const state = InternalStoreState.create([sliceDep1, sliceDep2, mySlice]);

    const newState = state.applyTransaction(mySlice.actions.myAction(5));

    const result = newState.getSliceState(mySlice);

    // @ts-expect-error - should error when a field is not defined
    let testVal0 = result.xyz;

    expectType<{ num: number } | undefined>(result);

    expect(newState.getSliceState(mySlice)).toEqual({
      num: 50 + 1 + 5 + 3,
    });
  });
});

describe('validations', () => {
  test('throws error if slice key not unique', () => {
    const mySlice = slice({
      key: createKey('test', [], { num: 1 }),
      actions: {
        myAction: (num: number) => (state) => {
          return { ...state, num: num + state.num };
        },
      },
    });

    const mySlice2 = slice({
      key: createKey('test', [], { num: 1 }),
      actions: {
        myAction: (num: number) => (state) => {
          return { ...state, num: num + state.num };
        },
      },
    });

    expect(() => {
      InternalStoreState.create([mySlice, mySlice2]);
    }).toThrowErrorMatchingInlineSnapshot(`"Duplicate slice keys key_test"`);
  });

  test('throws error if slice dependency is not registered', () => {
    const sliceDep = slice({
      key: createKey('test-dep', [], { num: 1 }),
      actions: {},
    });
    const mySlice = slice({
      key: createKey('test', [sliceDep], { num: 1 }),
      actions: {
        myAction: (num: number) => (state) => {
          return { ...state, num: num + state.num };
        },
      },
    });

    expect(() => {
      InternalStoreState.create([mySlice]);
    }).toThrowErrorMatchingInlineSnapshot(
      `"Slice "key_test" has a dependency on Slice "key_test-dep" which is either not registered or is registered after this slice."`,
    );
  });

  test('throws error if slice dependency is not registered before', () => {
    const sliceDep = slice({
      key: createKey('test-dep', [], { num: 1 }),
      actions: {},
    });
    const mySlice = slice({
      key: createKey('test', [sliceDep], { num: 1 }),
      actions: {
        myAction: (num: number) => (state) => {
          return { ...state, num: num + state.num };
        },
      },
    });

    expect(() => {
      InternalStoreState.create([mySlice, sliceDep]);
    }).toThrowErrorMatchingInlineSnapshot(
      `"Slice "key_test" has a dependency on Slice "key_test-dep" which is either not registered or is registered after this slice."`,
    );
  });
});

describe('test override helper', () => {
  test('overriding init state works', () => {
    const slice1 = slice({
      key: createKey('test1', [], { num: 1 }),
      actions: {},
    });

    const slice2 = slice({
      key: createKey('test2', [], { num: 2 }),
      actions: {},
    });

    let newState1 = InternalStoreState.create([
      slice1,
      testOverrideSlice(slice2, { initState: { num: 99 } }),
    ]);

    expect(newState1.getSliceState(slice1)).toEqual({ num: 1 });
    expect(newState1.getSliceState(slice2)).toEqual({ num: 99 });

    let newState2 = InternalStoreState.create([
      testOverrideSlice(slice1, { initState: { num: -1 } }),
      slice2,
    ]);
    expect(newState2.getSliceState(slice1)).toEqual({ num: -1 });
    expect(newState1.getSliceState(slice1)).toEqual({ num: 1 });
  });

  test('overriding effects works', () => {
    const slice1 = slice({
      key: createKey('test1', [], { num: 1 }),
      actions: {},
      effects: [
        {
          update: (sl) => {
            return;
          },
        },
      ],
    });

    expect(
      testOverrideSlice(slice1, { effects: [] }).spec.effects,
    ).toHaveLength(0);
    // should not affect initial slice
    expect(slice1.spec.effects).toHaveLength(1);
  });

  test('overriding dependencies', () => {
    const slice1 = slice({
      key: createKey('test1', [], { num: 1 }),
      actions: {},
    });

    expect(
      testOverrideSlice(slice1, { dependencies: [testSlice1] }).spec
        .dependencies.length,
    ).toBe(1);

    expect(slice1.spec.dependencies.length).toBe(0);
  });
});

describe('State creation', () => {
  test('empty slices', () => {
    const appState = InternalStoreState.create([]);

    expect(appState).toMatchInlineSnapshot(
      {
        _slices: expect.any(Array),
        slicesCurrentState: expect.any(Object),
        sliceLookupByKey: expect.any(Object),
      } as any,
      `
      {
        "_slices": Any<Array>,
        "context": undefined,
        "opts": undefined,
        "sliceLookupByKey": Any<Object>,
        "slicesCurrentState": Any<Object>,
      }
    `,
    );
  });

  test('with a slice', () => {
    const mySlice = slice({
      key: createKey('mySlice', [], { val: null }),
      actions: {},
    });

    const appState = InternalStoreState.create([mySlice]);

    expect(appState.getSliceState(mySlice)).toEqual({ val: null });
    expect(appState).toEqual({
      _slices: expect.any(Array),
      opts: undefined,
      sliceLookupByKey: expect.any(Object),
      slicesCurrentState: {
        key_mySlice: {
          val: null,
        },
      },
    });
  });

  test('throws error if action not found', () => {
    const mySlice = slice({
      key: createKey('mySlice', [], { val: null }),
      actions: {},
    });

    const appState = InternalStoreState.create([mySlice]);

    expect(() =>
      appState.applyTransaction(
        new Transaction({
          sourceSliceName: 'mySlice',
          sourceSliceKey: createSliceKey('mySlice'),
          payload: [5],
          actionId: 'updateNum',
        }),
      ),
    ).toThrowError(`Action "updateNum" not found in Slice "key_mySlice"`);
  });

  test('applying action preserves states of those who donot have apply', () => {
    const mySlice = slice({
      key: createKey('mySlice', [], { num: 0 }),
      actions: {},
    });

    const mySlice2 = slice({
      key: createKey('mySlice2', [mySlice], { num: 0 }),
      actions: {
        updateNum: (num: number) => (state, storeState) => {
          return { ...state, num };
        },
      },
    });

    const appState = InternalStoreState.create([mySlice, mySlice2]);
    expect(mySlice.getState(appState).num).toBe(0);

    let newAppState = appState.applyTransaction(mySlice2.actions.updateNum(4));
    expect(mySlice.getState(newAppState).num).toBe(0);
    expect(mySlice2.getState(newAppState).num).toBe(4);
  });

  test('applying action with selectors', () => {
    const key1 = createKey(
      'mySlice',
      [],
      {
        char: '1',
      },
      {
        s1: (state) => {
          return {
            val1_1: state.char,
          };
        },
      },
    );

    const mySlice1 = slice({
      key: key1,
      actions: {
        moreChar: (num: number) => (state) => {
          return {
            ...state,
            char: Array.from(
              {
                length: num,
              },
              () => state.char,
            ).join(''),
          };
        },
      },
    });

    const mySlice2 = slice({
      key: createKey(
        'mySlice2',
        [mySlice1],
        {
          char: '2',
        },
        {
          s2: (state, storeState) => {
            return {
              val2_1: mySlice1.resolveSelectors(storeState).s1,
              val2_2: state.char,
            };
          },
        },
      ),
      actions: {},
    });

    const mySlice3 = slice({
      key: createKey(
        'mySlice3',
        [mySlice1, mySlice2],
        { char: '3' },
        {
          s3: (state, storeState) => {
            return {
              val3_2: mySlice2.resolveSelectors(storeState).s2,
              val3_1: mySlice1.resolveSelectors(storeState).s1,
              val3_3: state.char,
            };
          },
        },
      ),
      actions: {},
    });

    const appState = InternalStoreState.create([mySlice1, mySlice2, mySlice3]);

    const result1 = {
      s3: {
        val3_1: {
          val1_1: '1',
        },
        val3_2: {
          val2_1: {
            val1_1: '1',
          },
          val2_2: '2',
        },
        val3_3: '3',
      },
    };
    expect(mySlice3.resolveSelectors(appState)).toEqual(result1);
    expect(mySlice2.resolveSelectors(appState).s2).toEqual(
      result1['s3']['val3_2'],
    );
    expect(mySlice1.resolveSelectors(appState).s1).toEqual(
      result1['s3']['val3_1'],
    );

    const newAppState = appState.applyTransaction(mySlice1.actions.moreChar(2));

    const result2 = {
      s3: {
        val3_1: {
          val1_1: '11',
        },
        val3_2: {
          val2_1: {
            val1_1: '11',
          },
          val2_2: '2',
        },
        val3_3: '3',
      },
    };

    expect(mySlice3.resolveSelectors(newAppState)).toEqual(result2);
    expect(mySlice2.resolveSelectors(newAppState).s2).toEqual(
      result2['s3']['val3_2'],
    );
    expect(mySlice1.resolveSelectors(newAppState).s1).toEqual(
      result2['s3']['val3_1'],
    );

    // previous state is unaltered
    expect(mySlice3.resolveSelectors(appState)).toEqual(result1);
  });
});

describe('Override init state', () => {
  const mySlice1 = slice({
    key: createKey('mySlice1', [], { num1: 0 }),
    actions: {},
  });
  const mySlice2 = slice({
    key: createKey('mySlice2', [], { num2: 0 }),
    actions: {},
  });
  test('can override state', () => {
    const appState = InternalStoreState.create([mySlice1, mySlice2], {
      mySlice1: { num1: 5 },
    });

    expect(mySlice1.getState(appState).num1).toBe(5);
  });

  test('throws error if slice not found', () => {
    expect(() =>
      InternalStoreState.create([mySlice1, mySlice2], {
        mySlice1: { num1: 5 },
        xSlice: { num1: 5 },
      }),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Some slice names (xSlice) in initStateOverride were not found in the provided slices"`,
    );
  });
});
