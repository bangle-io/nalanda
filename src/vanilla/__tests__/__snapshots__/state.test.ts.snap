// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`State creation with a slice 1`] = `
InternalStoreState {
  "_slices": [
    Slice {
      "_bare": {
        "keyMapping": KeyMapping {
          "_mapping": {
            "mySlice": "mySlice",
          },
        },
        "txApplicators": {},
        "txCreators": {},
      },
      "config": {
        "actions": {},
        "dependencies": [],
        "effects": [],
        "initState": {
          "val": null,
        },
        "key": "mySlice",
        "selectors": {},
      },
      "initState": {
        "val": null,
      },
      "key": "mySlice",
      "resolveSelectors": [Function],
      "resolveState": [Function],
    },
  ],
  "opts": undefined,
  "slicesCurrentState": {
    "mySlice": {
      "val": null,
    },
  },
}
`;
